pkg/operator/operator.go

```diff
@@ -321,108 +321,20 @@ func (optr *Operator) sync(key string) error {
-		return err
-	}
-	rootCA, err := optr.getCAsFromConfigMap("kube-system", "root-ca", "ca.crt")
-	if err != nil {
-		return err
-	}
-	// as described by the name this is essentially static, but it no worse than what was here before.  Since changes disrupt workloads
-	// and since must perfectly match what the installer creates, this is effectively frozen in time.
-	kubeAPIServerServingCABytes, err := optr.getCAsFromConfigMap("openshift-config", "initial-kube-apiserver-server-ca", "ca-bundle.crt")
-	if err != nil {
-		return err
-	}
-	bundle := make([]byte, 0)
-	bundle = append(bundle, rootCA...)
-	bundle = append(bundle, kubeAPIServerServingCABytes...)
-
-	// sync up os image url
-	// TODO: this should probably be part of the imgs
-	osimageurl, err := optr.getOsImageURL(namespace)
-	if err != nil {
-		return err
-	}
-	imgs.MachineOSContent = osimageurl
-
-	// sync up the ControllerConfigSpec
-	infra, network, proxy, err := optr.getGlobalConfig()
-	if err != nil {
-		return err
-	}
-	spec, err := createDiscoveredControllerConfigSpec(infra, network, proxy)
-	if err != nil {
-		return err
-	}
-
-	// if the cloudConfig is set in infra read the cloud config reference
-	if infra.Spec.CloudConfig.Name != "" {
-		cc, err := optr.getCloudConfigFromConfigMap("openshift-config", infra.Spec.CloudConfig.Name, infra.Spec.CloudConfig.Key)
-		if err != nil {
-			return err
-		}
-		spec.CloudProviderConfig = cc
-	}
-
-	spec.EtcdCAData = etcdCA
-	spec.EtcdMetricCAData = etcdMetricCA
-	spec.RootCAData = bundle
-	spec.PullSecret = &corev1.ObjectReference{Namespace: "openshift-config", Name: "pull-secret"}
-	spec.OSImageURL = imgs.MachineOSContent
-	spec.Images = map[string]string{
-		templatectrl.EtcdImageKey:            imgs.Etcd,
-		templatectrl.SetupEtcdEnvKey:         imgs.SetupEtcdEnv,
-		templatectrl.InfraImageKey:           imgs.InfraImage,
-		templatectrl.KubeClientAgentImageKey: imgs.KubeClientAgent,
-	}
-
-	// create renderConfig
-	rc := getRenderConfig(namespace, string(kubeAPIServerServingCABytes), spec, &imgs, infra.Status.APIServerURL)
 	// syncFuncs is the list of sync functions that are executed in order.
-	// any error marks sync as failure but continues to next syncFunc
+	// any error marks sync as failure but continues to next syncFunc unless it's the "render-config" one
+	// which is in charge of setting up the renderConfig for the subsequent syncfuncs.
 	var syncFuncs = []syncFunc{
+		// "render-config" must always run first as it sets the renderConfig in the operator
+		// for the sync funcs below. This is the only function that prevents the others from even running also.
+		{"render-config", optr.syncRenderConfig},
 		{"pools", optr.syncMachineConfigPools},
 		{"mcd", optr.syncMachineConfigDaemon},
 		{"mcc", optr.syncMachineConfigController},
 		{"mcs", optr.syncMachineConfigServer},
 		{"required-pools", optr.syncRequiredMachineConfigPools},
 	}
-	return optr.syncAll(rc, syncFuncs)
+	return optr.syncAll(optr.renderConfig, syncFuncs)
```

Comment:
Feels like this should really be:

```
if err := optr.SyncRenderConfig(); err != nil {
    return err
}
return optr.syncAll(syncFuncs)
```

And we use `optr.renderConfig` in the sync funcs or so?

Reply:
the thing is syncRenderConfig must follow operator versions settings as well, that's why it's in the array.

also, to reply to https://github.com/openshift/machine-config-operator/pull/911#discussion_r299654948 - it looks weird I know but we do already have functions taking renderConfig but not using it e.g.: `syncRequiredMachineConfigPools` and that's why we have the function signagture typed I guess

```
func (optr *Operator) syncRequiredMachineConfigPools(_ *renderConfig) error {
```

Reply:
also, wiring syncRenderConfig _outside_ of syncAll to respect all the Status flipping is really non-trivial (that's why I went this way)

Reply:
also, the other sync functions are gated on this first func to run anyway